src/System.Data.SqlClient/src/System/Data/Common/DataRecordInternal.cs

```diff
@@ -0,0 +1,323 @@
+                        cbytes = cbytes - ndataIndex;
+                    else
+                        cbytes = length;
+                }
+
+                // until arrays are 64 bit, we have to do these casts
+                Array.Copy(data, ndataIndex, buffer, bufferIndex, (int)cbytes);
+            }
+            catch (Exception e)
+            {
+                if (ADP.IsCatchableExceptionType(e))
+                {
+                    cbytes = data.Length;
+
+                    if (length < 0)
+                        throw ADP.InvalidDataLength(length);
+
+                    // if bad buffer index, throw
+                    if (bufferIndex < 0 || bufferIndex >= buffer.Length)
+                        throw ADP.InvalidDestinationBufferIndex(length, bufferIndex, "bufferIndex");
+
+                    // if bad data index, throw
+                    if (dataIndex < 0 || dataIndex >= cbytes)
+                        throw ADP.InvalidSourceBufferIndex(length, dataIndex, "dataIndex");
+
+                    // if there is not enough room in the buffer for data
+                    if (cbytes + bufferIndex > buffer.Length)
+                        throw ADP.InvalidBufferSizeOrIndex(cbytes, bufferIndex);
+                }
+
+                throw;
+            }
+
+            return cbytes;
+        }
+
+        public override char GetChar(int i)
+        {
+            string s;
+
+            s = (string)_values[i];
+            char[] c = s.ToCharArray();
+            return c[0];
+        }
+
+        public override long GetChars(int i, long dataIndex, char[] buffer, int bufferIndex, int length)
+        {
+            int cchars = 0;
+            string s;
+            int ndataIndex;
+
+            // if the object doesn't contain a char[] then the user will get an exception
+            s = (string)_values[i];
+
+            char[] data = s.ToCharArray();
+
+            cchars = data.Length;
+
+            // since arrays can't handle 64 bit values and this interface doesn't
+            // allow chunked access to data, a dataIndex outside the rang of Int32
+            // is invalid
+            if (dataIndex > Int32.MaxValue)
+            {
+                throw ADP.InvalidSourceBufferIndex(cchars, dataIndex, "dataIndex");
+            }
+
+            ndataIndex = (int)dataIndex;
+
+
+            // if no buffer is passed in, return the number of characters we have
+            if (null == buffer)
+                return cchars;
+
+            try
+            {
+                if (ndataIndex < cchars)
+                {
+                    // help the user out in the case where there's less data than requested
+                    if ((ndataIndex + length) > cchars)
+                        cchars = cchars - ndataIndex;
+                    else
+                        cchars = length;
+                }
+
+                Array.Copy(data, ndataIndex, buffer, bufferIndex, cchars);
```

Comment:
I think coping to char array could be avoided because `string` has method `CopyTo` which accepts `char[]`